### PR TITLE
Add app state to AppInfo

### PR DESCRIPTION
**Pull request by AFFogarty**
## What changes were proposed in this pull request?

This PR adds app state to `AppInfo`.  This change gives clients more power to understand the state of their application and to debug any failures.

For example, there is currently no way to differentiate between interactive sessions that have gone into YARN state `FAILED` or YARN state `FINISHED`. Livy reports both of these as `dead`.  With this new change, a client could inspect the `appState` to see the true state of the YARN application.

## How was this patch tested?

This change includes unit tests.

CC: @rapoth @imback82

### CODE REVIEW

File: server/src/main/scala/org/apache/livy/utils/SparkApp.scala [hunk 1]
```diff
@@ -19,17 +19,28 @@ package org.apache.livy.utils
...
+    var driverLogUrl: Option[String] = None,
+    var sparkUiUrl: Option[String] = None,
+    @JsonScalaEnumeration(classOf[StateTypeReference])
+    var appState: Option[SparkApp.State] = None) {
   import AppInfo._
   def asJavaMap: java.util.Map[String, String] =
-    Map(DRIVER_LOG_URL_NAME -> driverLogUrl.orNull, SPARK_UI_URL_NAME -> sparkUiUrl.orNull).asJava
+    Map(DRIVER_LOG_URL_NAME -> driverLogUrl.orNull,
+      SPARK_UI_URL_NAME -> sparkUiUrl.orNull,
+      APP_STATE_NAME -> appState.map(s => s.toString).orNull).asJava
```
Comment (imback82) [inline server/src/main/scala/org/apache/livy/utils/SparkApp.scala:43]: nit: `appState.map(_.toString)`

File: server/src/test/scala/org/apache/livy/server/interactive/InteractiveSessionServletSpec.scala [hunk 2]
```diff
@@ -168,7 +168,10 @@ class InteractiveSessionServletSpec extends BaseInteractiveServletSpec {
     val proxyUser = "proxyUser"
     val state = SessionState.Running
     val kind = Spark
-    val appInfo = AppInfo(Some("DRIVER LOG URL"), Some("SPARK UI URL"))
+    val appInfo = AppInfo(
+      Some("DRIVER LOG URL"),
+      Some("SPARK UI URL"),
+      Some(SparkApp.State.RUNNING))
```
Comment (imback82) [inline server/src/test/scala/org/apache/livy/server/interactive/InteractiveSessionServletSpec.scala:174]: can we also test `None` case?

Comment (AFFogarty) [reply server/src/test/scala/org/apache/livy/server/interactive/InteractiveSessionServletSpec.scala:174]: Added verification that `appState=None` gets serialized to `"appState": null`.

**Comment by imback82** [comment]
@jerryshao Could you help reviewing this PR? Thanks in advance!

File: server/src/main/scala/org/apache/livy/utils/SparkApp.scala (hunk 1 above)
Comment (AFFogarty) [reply server/src/main/scala/org/apache/livy/utils/SparkApp.scala]: Fixed

**Comment by andrasbeni** [comment]
@AFFogarty, I'm curious what you think of the following alternatives:
- Mapping SparkApp.State.FAILED to SessionState.ERROR()
- Introducing SessionState.FAILED()
I'm not too familiar with the state transitions, so above ideas may be dumb. In which case I'm happy to learn why.

**Comment by rapoth** [comment]
@jerryshao Any chance you could review this PR? Thanks a lot for your time in advance!
